Force docking

Thaums opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

Replaces the autopilot console with an "Auxiliary Console" that is basically the same but comes with an additional mode that lets it use a new machine called the Interdictor
2 New things:
Interdictor: Use it to forcibly dock a nearby ship. There's 2 ways to use it: The easiest way is to link it to the Auxiliary Console and use it through that. The second option is to find the wire that triggers its subvert function. If you do it this way then it'll automatically target the nearest (eligible) ship.
Antivirus: Load it into your helm. If someone tries to subvert your ship it will block the attempt and use up the antivirus. It will also tell you which ship tried to subvert you.

Getting a bit more in-depth
- The interdictor has built in checks for the factions to prevent NEU fuckery
- You can research space hacking to print a interdictor board or order one via cargo
- interdictors have a cooldown which can be reduced with better parts, the parts do other things too (like how long your victim's helm controls are locked)
- After recovering from a subversion the ship will have a 1 minute grace period where they cannot be subverted
- Antiviruses can only be bought from cargo (under Security)
- Some other stuff, like interdictors causing the room to heat up if you use them too much

Supernova wanted a "request docking" feature, thats not in this because I accidently erased all the files and had to start over, and I've been working on this for nearly a week

I have tested it, haven't noticed any issues (though I'm sure there are some)

## Why It's Good For The Game

The fact that you can just press a button and become immune to assault from other ships isn't that great, bunch of people want this and the faction checks mean NEU players shouldnt be bothered

## Changelog
:cl:
add: Added new machine Interdictor, use it to forcibly drag ships out of hyperspace
add: Antivirus, prevent the above from happening
edit: Autopilot console repurposed as Auxiliary console, retains the same function as before
/:cl:
![Capture](https://user-images.githubusercontent.com/21028062/155231670-1976cfb1-2fe9-4e31-9c9b-18c5fc7dde4a.PNG)


<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

Thaums committed:
Update bedsheet_bin.dm

Thaums committed:
Merge branch 'master' of https://github.com/Jackriip/Voidcrew-LRP

Thaums committed:
Merge branch 'master' of https://github.com/Jackriip/Voidcrew-LRP

Thaums committed:
Merge branch 'master' of https://github.com/Jackriip/Voidcrew-LRP

Thaums committed:
Merge branch 'master' of https://github.com/voidcrew/Voidcrew-LRP

Thaums committed:
Merge branch 'master' of https://github.com/voidcrew/Voidcrew-LRP

Thaums committed:
Merge branch 'master' of https://github.com/voidcrew/Voidcrew-LRP

Thaums committed:
Merge branch 'master' of https://github.com/voidcrew/Voidcrew-LRP

Thaums committed:
Merge branch 'voidcrew:master' into master

Thaums committed:
Merge branch 'master' of https://github.com/thaums/Voidcrew-LRP

Thaums committed:
force docking

Thaums committed:
spelling mistakes

its fucking auxiliary you dumbass

SuperNovaa41 commented on voidcrew/code/game/machinery/computer/shuttle/auto_pilot.dm:
```diff
@@ -20,20 +28,38 @@
 		return TRUE
 
+/obj/machinery/computer/autopilot/attackby(obj/item/O, mob/user, params)
+	if(istype(O, /obj/item/multitool))
```
just use `multitool_act`

SuperNovaa41 commented on voidcrew/code/game/machinery/computer/shuttle/auto_pilot.dm:
```diff
@@ -74,6 +100,45 @@
+		Subverter ui data
+	*/
+
+	data["subverter"] = sub
```
do you want to assign it even if its null?

Thaums replied:
Im not sure what the issue is here, if sub is null then none of its fields are accessed and the tgui disables the subverter mode. I'm fairly confident that i want to assign it like that

SuperNovaa41 commented on voidcrew/code/game/machinery/computer/shuttle/auto_pilot.dm:
```diff
@@ -74,6 +100,45 @@
+			for(var/location in data["ships"])
+				subvert_target = location["id"]
+				data["selected_target"] = subvert_target
+		if(!length(data["ships"]))
```
this should be else if, otherwise theres a chance for some weird edge cases

SuperNovaa41 commented on voidcrew/code/modules/overmap/ships/pvp/antivirus.dm:
```diff
@@ -0,0 +1,15 @@
+	desc = "A disk containing a single antiviral node, good for preventing one ship subversion. Load into helm."
+	random_color = FALSE
+	color = "#FFFF00"
+	custom_materials = list(/datum/material/iron=300, /datum/material/glass=100)
```
```suggestion
	custom_materials = list(/datum/material/iron = 300, /datum/material/glass = 100)
```
bike shedding but i hate non spaced equal signs

Thaums committed:
by popular demand

changes name to interdictor, a reference to a game ive never played

Thaums committed:
missed a spot

SuperNovaa41 commented on voidcrew/code/modules/overmap/ships/pvp/misc.dm:
```diff
@@ -0,0 +1,55 @@
+#define SHIP_GRACE_TIME (1 MINUTES)
```
make sure to `#undef` this at the end of the file (i assume it's used nowhere else)

SuperNovaa41 commented on voidcrew/code/modules/overmap/ships/pvp/misc.dm:
```diff
@@ -0,0 +1,55 @@
+
+/obj/structure/overmap/ship/simulated
+	var/obj/machinery/computer/helm/most_recent_helm // For sending messages to the target's helm
+	var/engine_cooldown = 0 // The point in time when your engines will work again
```
you should use the cooldown macros for this (it makes managing cooldowns far easier), see `code/__DEFINES/cooldowns.dm` (at the bottom) for more details

SuperNovaa41 commented on voidcrew/code/modules/overmap/ships/pvp/misc.dm:
```diff
@@ -0,0 +1,55 @@
+
+/obj/machinery/computer/helm/connect_to_shuttle(obj/docking_port/mobile/port, obj/docking_port/stationary/dock)
+	..()
+	current_ship.most_recent_helm = src
```
this proc already exists in `voidcrew/code/modules/overmap/helm.dm` you should just put it there

SuperNovaa41 commented on voidcrew/code/modules/overmap/ships/pvp/subverter.dm:
```diff
@@ -0,0 +1,235 @@
+#define SUBVERTER_RECHARGE_TIME (2 MINUTES)
+#define SUBVERTER_ENGINE_STALL_TIME (1 MINUTES)
+
+#define SUBVERTER_SPICINESS 150
```
make sure to `#undef` these

SuperNovaa41 commented on voidcrew/code/modules/overmap/ships/pvp/subverter.dm:
```diff
@@ -0,0 +1,235 @@
+	var/disabled = FALSE
+	var/shocked = FALSE
+	//TODO: keep track of which map the subverter is on
+	var/subverter_cooldown = 0
```
cooldown macro here

SuperNovaa41 commented on voidcrew/code/modules/overmap/ships/pvp/subverter.dm:
```diff
@@ -0,0 +1,235 @@
+
+//quick maths
+/obj/machinery/subverter/proc/can_subvert(obj/structure/overmap/ship/simulated/target_ship, antag = FALSE)
+	return !(((!antag) && (target_ship.prefix == "NEU" || (ship && (ship.prefix == "NEU" || (ship.prefix != "KOS" && ship.prefix == target_ship.prefix))))) || world.time < subverter_cooldown || world.time < target_ship.engine_cooldown || world.time < target_ship.sub_grace || target_ship.state != OVERMAP_SHIP_FLYING)
```
this is really gross, you should really try to avoid long boolean statements like this because they extremely hard to read.

Heres an idea:
use this proc to return an int for each different fail case, and in the above proc (`bark_processing`) switch through the returned int, and give the error message then, so that you don't have to check for the exact same errors in two different procs.
You should define all the ints of course, so that its understandable what errors are being sent back

Thaums replied:
i was gonna say what if instead of a switch case, theres just a list that stores all the strings and the int indexes through it but that might be too much effort to also have that keep track of what sound to play

SuperNovaa41 commented on voidcrew/code/modules/overmap/ships/pvp/misc.dm:
```diff
@@ -0,0 +1,55 @@
+#define SHIP_GRACE_TIME (1 MINUTES)
+
+/obj/structure/overmap/ship/simulated
+	var/obj/machinery/computer/helm/most_recent_helm // For sending messages to the target's helm
```
why is it called `most_recent_helm`? ships can only have one helm

SuperNovaa41 commented on voidcrew/code/modules/overmap/ships/pvp/subverter.dm:
```diff
@@ -0,0 +1,235 @@
+/obj/machinery/subverter/proc/attempt_to_subvert(obj/structure/overmap/ship/simulated/target_ship)
+	. = FALSE
+	bark_processing(target_ship)
+	if (can_subvert(target_ship))
```
early return here
```suggestion
	if (!can_subvert(target_ship))
	    return
	...
```

SuperNovaa41 commented on voidcrew/code/modules/overmap/ships/pvp/subverter.dm:
```diff
@@ -0,0 +1,235 @@
+	update_stats()
+
+/obj/machinery/subverter/proc/its_gettin_hot_in_here()
+	//var/turf/T = get_turf(src)
```
?

SuperNovaa41 commented on voidcrew/code/modules/overmap/ships/pvp/subverter.dm:
```diff
@@ -0,0 +1,235 @@
+
+/obj/machinery/subverter/proc/its_gettin_hot_in_here()
+	//var/turf/T = get_turf(src)
+	var/turf/L = loc
```
loc isn't always guaranteed to be a turf, you should definitely use the commented out code above

SuperNovaa41 commented on voidcrew/code/modules/overmap/ships/pvp/subverter.dm:
```diff
@@ -0,0 +1,235 @@
+*/
+/obj/machinery/subverter/proc/adjust_hacked(state)
+	hacked = state
+	if (hacked)
```
early return here

SuperNovaa41 commented on voidcrew/code/modules/overmap/ships/pvp/subverter.dm:
```diff
@@ -0,0 +1,235 @@
+		wires.interact(user)
+		return TRUE
+
+	if(istype(O, /obj/item/multitool))
```
multitool act

SuperNovaa41 commented on voidcrew/code/modules/overmap/ships/pvp/subverter_circuit.dm:
```diff
@@ -0,0 +1,28 @@
+	name = "Interdictor (Machine Board)"
+	icon_state = "engineering"
+	build_path = /obj/machinery/subverter
+	custom_materials = list(/datum/material/glass=1000, /datum/material/bluespace=8000)
```
```suggestion
	custom_materials = list(/datum/material/glass = 1000, /datum/material/bluespace = 8000)
```

SuperNovaa41 commented on voidcrew/code/modules/overmap/ships/pvp/subverter_circuit.dm:
```diff
@@ -0,0 +1,28 @@
+	name = "Machine Design (Interdictor)"
+	desc = "The circuit board for an interdictor."
+	id = "shuttle_subverter"
+	materials = list(/datum/material/glass=1000, /datum/material/bluespace=8000)
```
```suggestion
	materials = list(/datum/material/glass = 1000, /datum/material/bluespace = 8000)
```

SuperNovaa41 commented on whitesands/code/modules/research/techweb/all_nodes.dm:
```diff
@@ -13,7 +13,7 @@
 	description = "Research the technology required to create and use basic shuttles."
 	prereq_ids = list("bluespace_travel", "adv_engi")
-	design_ids = list("engine_plasma", "engine_ion", "engine_heater", "engine_smes", "shuttle_helm") //, "rapid_shuttle_designator")
+	design_ids = list("engine_plasma", "engine_ion", "engine_heater", "engine_smes", "shuttle_helm", "shuttle_autopilot") //, "rapid_shuttle_designator")
```
hey while you're here why don't you just move all of the `shuttle_tech` research nodes to voidcrew

SuperNovaa41 reviewed (changes requested):
All of your new procs, and new variables need to be properly documented.
```
///comment describing the variable
var/new_variable

/**
 * Comment describing the proc
 *
 * Arguments:
 * /mob/living/carbon/human/user - The user
 * example - An example argument
 */
/proc/new_proc(mob/living/carbon/human/user, example)
    ...
```

also you shouldn't have ANY single letter variables, they should all be named related to whatever their purpose is (var/turf/turf, var/mob/mob, etc is acceptable)

Thaums committed:
Update voidcrew/code/modules/overmap/ships/pvp/antivirus.dm

Co-authored-by: Seth Scherer <supernovaa41@gmx.com>

Thaums committed:
Update voidcrew/code/modules/overmap/ships/pvp/subverter_circuit.dm

Co-authored-by: Seth Scherer <supernovaa41@gmx.com>

Thaums committed:
Update voidcrew/code/modules/overmap/ships/pvp/subverter_circuit.dm

Co-authored-by: Seth Scherer <supernovaa41@gmx.com>

Thaums committed:
requested changes

documented new procs and vars
used cooldown macros
using multitool_act
removed one letter vars
some other stuff i cant remember

SuperNovaa41 commented on voidcrew/code/game/machinery/computer/shuttle/auto_pilot.dm:
```diff
@@ -19,22 +27,54 @@
 		ship = port.current_ship
 		return TRUE
+/*
+/obj/machinery/computer/autopilot/attackby(obj/item/O, mob/user, params)
```
Why did you comment this out? just remove it

Thaums replied:
forgot to remove that, left over from replacing it with multitool_act

SuperNovaa41 commented on voidcrew/code/modules/overmap/ships/pvp/subverter.dm:
```diff
@@ -0,0 +1,329 @@
+*	obj/structure/overmap/ship/simulated/target_ship - The target ship of the subversion attempt
+*	bark_success_message - (optional) Text displayed upon a successful subversion
+*/
+/obj/machinery/subverter/proc/bark_processing(subvert_flag, obj/structure/overmap/ship/simulated/target_ship, bark_success_message = "Agent was successful! Bringing [target_ship.name] out of hyperspace.")
```
you don't need `target_ship` anymore

Thaums replied:
i do actually need that for the success message but im just gonna edit it so it just says "target vessel" like the other messages

SuperNovaa41 replied:
oh i didn't see it in the message whoops

SuperNovaa41 commented on voidcrew/code/modules/overmap/ships/pvp/subverter.dm:
```diff
@@ -0,0 +1,329 @@
+	if (aux)
+		aux.sub = null
+		aux = null
+	..()
```
```suggestion
	return ..()
```
looks better this way

SuperNovaa41 commented on voidcrew/code/modules/overmap/ships/pvp/subverter.dm:
```diff
@@ -0,0 +1,329 @@
+*/
+/obj/machinery/subverter/proc/get_recharge_str()
+	. = COOLDOWN_TIMELEFT(src, subverter_cooldown) / 10
+	. *= . > 0
```
Whats the point of this?
if `.` is 0, then multiply it by zero, otherwise multiply by one? this doesn't seem necessary

Thaums replied:
oh ye thats leftover from when i didnt use the macros

SuperNovaa41 commented on voidcrew/code/modules/overmap/ships/pvp/subverter.dm:
```diff
@@ -0,0 +1,329 @@
+	spark.start()
+	if (electrocute_mob(user, get_area(src), src, 0.7, TRUE))
+		return TRUE
+	else
```
this else is redundant

Thaums replied:
lol i just copied that from the autholathe.dm

SuperNovaa41 reviewed (changes requested):
This looks way better, just a couple thing, any new vars should have three slashes `///`, it's for code docs, you're also missing documentation on a couple vars, things like `spiciness`.. WTF is that??? (i would know if you documented that)

Thaums committed:
more fixes

Thaums committed:
FUCK

Thaums committed:
Fixes docking bug, adds request docking

Note that to fix this bug and properly implement request docking, some modifications had to be made to dynamic_events.dm and simulated.dm, which I don't like doing but seemed necessary. Hope it doesnt interfere with the rebase too much

Thaums committed:
Merge branch 'voidcrew:master' into master

Thaums committed:
thanks nova

Thaums committed:
Merge branch 'voidcrew:master' into master

SuperNovaa41 commented on tgui/packages/tgui/interfaces/HelmConsole.js:
```diff
@@ -168,9 +168,29 @@ const ShipContent = (props, context) => {
   return (
     <>
+  	  {!!dock_request && (
+      <Section title = "Docking Request">
```
weird indentation here

SuperNovaa41 commented on voidcrew/code/modules/overmap/dynamic_object_defines.dm:
```diff
@@ -14,6 +14,9 @@
 	///What kind of planet the level is, if it's a planet at all.
 	var/datum/overmap/planet/planet
+	///Keep track of whether or not the docks have been reserved by a ship. This is required to prevent issues where two ships will attempt to dock in the same place due to unfortunate timing
+	var/first_dock_taken = FALSE
+	var/second_dock_taken = FALSE
```
couldn't this just be done as an int? and then you check like `if (used_docks >= 2) ... dont dock`

Thaums replied:
Thats going to make it a lot harder to read, since the second dock could be taken but not the first. Its doable but im gonna need to define a bunch of ints and I think if we ever want to add more than 2 docks in the future then that approach is gonna make it even more confusing

SuperNovaa41 commented on voidcrew/code/modules/overmap/ships/pvp/misc.dm:
```diff
@@ -0,0 +1,119 @@
+	if (acting.state != OVERMAP_SHIP_FLYING)
+		acting.most_recent_helm.say("Must be in hyperspace to request docking.")
+		return
+	/*if (!is_still() || !acting.is_still())
```
why is this commented out

Thaums replied:
I left a comment sort of explaining but i put in a place thats hard to see cause im retarded, basically I'm not sure if that should be a condition because I don't think it would cause any issues to not have it, and it would just make trying to request a dock more annoying. If necessary it can be uncommented though

Thaums replied:
Actually I'm going to just put in a check when the request gets accepted to make sure both ships are on the same coordinate, that should be fine

SuperNovaa41 commented on voidcrew/code/modules/overmap/ships/pvp/subverter.dm:
```diff
@@ -0,0 +1,345 @@
+		return
+	COOLDOWN_START(src, subverter_cooldown, sub_recharge)
+	its_gettin_hot_in_here()
+	//use_power(active_power_usage)
```
?

SuperNovaa41 commented on voidcrew/code/modules/overmap/ships/pvp/subverter.dm:
```diff
@@ -0,0 +1,345 @@
+	//find nearest non-allied pvp ship and attempt to subvert
+	var/obj/structure/overmap/ship/simulated/nearest
+	var/ship_min_dist
+	var/can_sub_flag
```
define this in the for loop it's not used anywhere else

SuperNovaa41 commented on voidcrew/code/modules/overmap/ships/pvp/subverter.dm:
```diff
@@ -0,0 +1,345 @@
+		return
+	//find nearest non-allied pvp ship and attempt to subvert
+	var/obj/structure/overmap/ship/simulated/nearest
+	var/ship_min_dist
```
same here

Thaums committed:
requested changes

Thaums committed:
Merge branch 'voidcrew:master' into master
